### DISCUSSION TITLE
Support value tuples with single value using an extra comma in the list (example `var tuple = (int x,)`)

### DISCUSSION THREAD

**Discussion by ja72** [dotnet/csharplang#4668] [Language Ideas]
Currently `C#` supports `ValueTuple<T>` but not the syntax

```cs
var idata = (1);
var sdata = ("a");
```

because obviously the above is valid syntax for expressions.

I propose to allow single-valued tuples in a similar syntax to `Julia` with a comma after the _one_ element.

```cs
var idata = (1,);
var sdata = ("a",);
```

The intent is to be able to have a _named_ variable in the `ValueTuple<T>` that isn't `Item1`

```cs
class Foo1
{
    internal (int x,) data;
}

int Main()
{
    var foo = new Foo();
    foo.data.x = 10;
    // instead of foo.data.Item1 = 10;
}
```

and for consistency allow a comma at the end of the tuple value in general

```cs
var point = (1,2,3,);
var letters = ("a","b",);
```
Of course, this might cause visual confusion with discarded values when `(1,2,3,_)` means there are 4 elements in the tuple and `(1,2,3,)` would mean there is 3 elements in the tuple.

**Comment by CyrusNajmabadi** [comment]
Dupe of: https://github.com/dotnet/csharplang/issues/1595

**Reply by ja72** [reply]
It is not a dupe because I am actually proposing a syntax feature that circumvents the reason #1595 was closed. The added comma syntax isn't that far off for `C#` since it is used widely anytime there are lists defined. For example,

```cs
    public enum TriangleVertex
    {
        A,
        B,
        C,
    }
```

or

```cs
var list = new[] { 1, 2, 3, };
```

so why not

```cs
var tuple1 =(1,);
var tuple2 =(1,2,);
var tuple3 =(1,2,3,);
```

**Reply by CyrusNajmabadi** [reply]
See https://github.com/dotnet/csharplang/issues/575

This is all in the space of supporting womples in some way.

**Comment by Shadowblitz16** [comment]
Please don't add awful syntax just because `(0)` is already valid in C#.
It's ugly and we can just check of the expression inside the braces results in a value, if it's a logical expression then treat it that way.

**Comment by gerneio** [comment]
I was needing this today. My personal take is I wanted a named single value tuple so that I can easily expand the return type of a method in the future when/if I need to return additional data points. For example:

`List<int> GetItems(); // Intent not clear what "int" returns here`
vs
`List<(int ID,)> GetItems(); // Intent very clear what "int" means for this return`

So, traditionally, with the first method, to explicitly imply that I am returning a method with ID's, I would typically name it something like "GetItemIDs", however if my future intent is to also return other properties of Item, then I'd have to create a whole new method, or modify the function and update every consumer. What I would like to do is specify upfront that I am returning a complex type, that at least has an ID property, which can easily be appended to down-the-road w/o breaking any downstream consumers (assuming I don't change type/property name).

Of course, by default I could just always return a struct/class, however the point here is sometimes you just want to quickly scaffold up a method that you know will have a complex return type down-the-road, but you're unsure what structure that object will take atm. Currently I am doing the below work around, which IMO is much much uglier than the OP's suggested ending-comma approach.

`List<(int ID, object _)> GetItems();`

**Reply by DavidArno** [reply]
```cs
record Item(int id);

List<Item> GetItems();
```
"*so that I can easily expand the return type of a method in the future*". This is pure YAGNI in my view. But if you really want to "futureproof" your code, records already exist as an easy way of achieving this without the need for ugly syntax or "workarounds".

**Reply by daiplusplus** [reply]
Using a `record class` introduces significant overhead (it's a reference-type) if all you you want to accomplish is to add names to return-values or type-parameters: `List<Int32>` vs `List<SomeRecordType>` vs `List<( Int32 userId )>`. Defining a custom `record struct` is an option but still makes the DX clunky and _might_ introduce overheads of its own.

**Reply by CyrusNajmabadi** [reply]
@daiplusplus you'd get the same with a single-value-tuple.  It would be a struct after all.  Tuples are effectively just an anonymous record-struct.  So if you just want structural types (but a minimum of 2 elements), then a tuple is good.  If you want 1 element, and are ok with a nominal type, then a record struct works.

**Reply by daiplusplus** [reply]
@CyrusNajmabadi It'd be nice if a `record struct Foo( T x )` had an automatic `operator implicit T` defined 👀

**Reply by CyrusNajmabadi** [reply]
you could always create a source-generator that does exactly that for you :)

i don't think there's any appetite for doing that by default.  Especially as the constructor may throw, and implicit operators should not do that.

**Reply by Richiban** [reply]
> i don't think there's any appetite for doing that by default. Especially as the constructor may throw, and implicit operators should not do that.

I think daiplusplus is talking about the implicit conversion going the other way, from `Wrapper<T> -> T`. I agree that there shouldn't ever really be an implicit conversion `T -> Wrapper<T>`.

**Reply by CyrusNajmabadi** [reply]
Ah, sorry for the misunderstanding.  Thanks for the clarification!

**Comment by ja72** [comment]
Not everyone is familiar with source generators. Maybe a quick example is in order here.

**Reply by CyrusNajmabadi** [reply]
@ja72 a source generator just writes it a regular file. So you just need to write out a file containing another partial part for the record that contains the operator.

**Comment by jez9999** [comment]
Sigh, I wish C# had this.  I don't know why it wasn't put in in the first place.  I have methods returning 5, 4, 3, or 2 values and I can name them nicely.  But if I decide I want to remove one, bam I'm down to one return value and I have to totally switch to having an unnamed return value.  Please implement this feature instead of more pointless stuff like "top level statements" that adds nothing of value.

**Reply by AndersKehlet** [reply]
Poopooing previous work they've done is not how you convince people to cater to your preferences.

**Reply by jez9999** [reply]
Bit touchy are we?

**Reply by CyrusNajmabadi** [reply]
@AndersKehlet  @jez9999 Please keep the discussion on topic.  Thanks!

**Reply by bernd5** [reply]
Why you don't return the value directly - e.g. an int?

**Reply by CyrusNajmabadi** [reply]
@bernd5 He's saying he would like that value to have a name.   Whether or not that is important enough to warrant a lang feature remains to be seen :)

**Reply by jez9999** [reply]
If it's not that important, why did C# bother having names for 2, 3, 4, etc. element tuples?

**Reply by HaloFour** [reply]
> If it's not that important, why did C# bother having names for 2, 3, 4, etc. element tuples?

To make it easier to extract the correct data elements from the container when there is more than one of them.  Otherwise it's relatively easy to accidentally transpose the indexes, especially if the elements are of the same data type.

**Reply by jez9999** [reply]
Make it easier?  Well.  Doesn't sound necessary, does it?  Only convenient.  Are convenience features worth implementing?

**Reply by HaloFour** [reply]
@jez9999 

> Are convenience features worth implementing?

There is never a blanket answer to that statement.

With 2+ arity tuples the specific concern was transposing the index, which would lead to subtle logic errors.  Naming the elements neatly solves that problem.  So, in this case, it was considered worthwhile.  That problem doesn't exist for single return values or "womples".

**Reply by CyrusNajmabadi** [reply]
> why did C# bother having names for 2, 3, 4, etc. element tuples?

Because in those cases having names to indicate meaning was more important.
